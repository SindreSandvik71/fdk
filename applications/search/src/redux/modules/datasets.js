import _ from 'lodash';
import { addOrReplaceParam } from '../../lib/addOrReplaceUrlParam';
import { fetchActions } from '../fetchActions';

export const DATASETS_REQUEST = 'DATASETS_REQUEST';
export const DATASETS_SUCCESS = 'DATASETS_SUCCESS';
export const DATASETS_FAILURE = 'DATASETS_FAILURE';

function createNestedListOfPublishers(listOfPublishers) {
  const nestedListOfPublishers = _(listOfPublishers).forEach(publisherItem => {
    const filteredChildrenOfParentPublishers = _(listOfPublishers)
      .filter(
        g => g.key.substring(0, g.key.lastIndexOf('/')) === publisherItem.key
      )
      .value();

    filteredChildrenOfParentPublishers.forEach(item => {
      const retVal = item;
      retVal.hasParent = true;
      return retVal;
    });

    const retVal = publisherItem;
    retVal.children = filteredChildrenOfParentPublishers;
    return retVal;
  });

  const resultArray = _(nestedListOfPublishers)
    .filter(f => !f.hasParent)
    .value();

  return resultArray;
}

export function fetchDatasetsIfNeededAction(datasetsURL) {
  // add static size parameter
  const url = addOrReplaceParam(datasetsURL, 'size', '50');
  return (dispatch, getState) => {
    if (!getState().datasets.isFetching) {
      dispatch(
        fetchActions(url, [
          DATASETS_REQUEST,
          DATASETS_SUCCESS,
          DATASETS_FAILURE
        ])
      );
    }
  };
}

const initialState = {
  isFetching: false,
  datasetItems: null,
  publisherCountItems: null
};

export function datasetsReducer(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST: {
      return {
        ...state,
        isFetching: true
      };
    }
    case DATASETS_SUCCESS: {
      const objFromArray = action.payload.aggregations.subjectsCount.buckets.reduce(
        (accumulator, current) => {
          accumulator[current.key] = current; // eslint-disable-line no-param-reassign
          return accumulator;
        },
        {}
      );
      return {
        ...state,
        isFetching: false,
        datasetItems: action.payload,
        publisherCountItems: createNestedListOfPublishers(
          action.payload.aggregations.orgPath.buckets
        ),
        subjectsCountItems: objFromArray
      };
    }
    case DATASETS_FAILURE: {
      return {
        ...state,
        isFetching: false,
        datasetItems: null,
        publisherCountItems: null
      };
    }
    default:
      return state;
  }
}

export const getDatasetById = (datasets, id) => {
  if (datasets && datasets.datasetItems && datasets.datasetItems.hits) {
    return datasets.datasetItems.hits.hits.filter(
      dataset => dataset._source.id === id
    );
  }
  return null;
};
