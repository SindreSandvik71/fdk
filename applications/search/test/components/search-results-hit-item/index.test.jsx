import React from 'react';
import { shallow } from 'enzyme';
import SearchHitItem from '../../../src/components/search-results-hit-item';
import datasets from '../../fixtures/datasets';

let defaultProps;

test('should render SearchHitItem NON_PUBLIC correctly', () => {
  defaultProps = {
    result: datasets[0]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render SearchHitItem RESTRICTED correctly', () => {
  defaultProps = {
    result: datasets[1]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render SearchHitItem PUBLIC correctly', () => {
  defaultProps = {
    result: datasets[2]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render SearchHitItem with no title, description, accessRights and distribution with no format correctly', () => {
  defaultProps = {
    result: datasets[3]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
