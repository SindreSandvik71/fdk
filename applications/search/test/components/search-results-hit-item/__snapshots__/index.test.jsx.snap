// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchHitItem NON_PUBLIC correctly 1`] = `
<Link
  className="fdk-a-search-hit"
  replace={false}
  title="Datasett: AAMarkagrensen Oslo Kommune og nærliggende kommuner"
  to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
  <div
    className="fdk-container-search-hit"
  >
    <h2>
      AAMarkagrensen Oslo Kommune og nærliggende kommuner
    </h2>
    <div
      className="fdk-dataset-themes"
    >
      <span>
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettet
        </span>
        Eies av
         
        <span
          className="fdk-strong-virksomhet"
        >
          Testetaten
        </span>
      </span>
      <div
        className="fdk-label mr-2 mb-2"
        key="dataset-description-theme-0"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettets tema.
        </span>
        Forvaltning og offentlig støtte
      </div>
      <div
        className="fdk-label mr-2 mb-2"
        key="dataset-description-theme-1"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettets tema.
        </span>
        Miljø
      </div>
    </div>
    <p
      className="fdk-p-search-hit"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Beskrivelse av datasettet,
      </span>
      Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4...
    </p>
    <div
      className="fdk-container-distributions fdk-distributions-red"
    >
      <strong>
        Unntatt offentlighet
      </strong>
      <div />
      <div
        id="search-hit-sample"
      >
        Eksempeldata tilgjengelig
      </div>
    </div>
  </div>
</Link>
`;

exports[`should render SearchHitItem PUBLIC correctly 1`] = `
<Link
  className="fdk-a-search-hit"
  replace={false}
  title="Datasett: AAMarkagrensen Oslo Kommune og nærliggende kommuner"
  to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
  <div
    className="fdk-container-search-hit"
  >
    <h2>
      AAMarkagrensen Oslo Kommune og nærliggende kommuner
    </h2>
    <div
      className="fdk-dataset-themes"
    >
      <div
        className="fdk-label mr-2 mb-2"
        key="dataset-description-theme-0"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettets tema.
        </span>
        Forvaltning og offentlig støtte
      </div>
      <div
        className="fdk-label mr-2 mb-2"
        key="dataset-description-theme-1"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettets tema.
        </span>
        Miljø
      </div>
    </div>
    <p
      className="fdk-p-search-hit"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Beskrivelse av datasettet,
      </span>
      Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder Datasettes formål er nullam quis rius eget urna mollis ornare vel eu leo. Praesent commo ...
    </p>
    <div
      className="fdk-container-distributions fdk-distributions-green"
    >
      <strong>
        Offentlig
      </strong>
    </div>
  </div>
</Link>
`;

exports[`should render SearchHitItem RESTRICTED correctly 1`] = `
<Link
  className="fdk-a-search-hit"
  replace={false}
  title="Datasett: AAMarkagrensen Oslo Kommune og nærliggende kommuner"
  to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
  <div
    className="fdk-container-search-hit"
  >
    <h2>
      AAMarkagrensen Oslo Kommune og nærliggende kommuner
    </h2>
    <div
      className="fdk-dataset-themes"
    >
      <span>
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettet
        </span>
        Eies av
         
        <span
          className="fdk-strong-virksomhet"
        >
          Testetaten
        </span>
      </span>
      <div
        className="fdk-label mr-2 mb-2"
        key="dataset-description-theme-0"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettets tema.
        </span>
        Forvaltning og offentlig støtte
      </div>
      <div
        className="fdk-label mr-2 mb-2"
        key="dataset-description-theme-1"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettets tema.
        </span>
        Miljø
      </div>
    </div>
    <p
      className="fdk-p-search-hit"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Beskrivelse av datasettet,
      </span>
      Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4...
    </p>
    <div
      className="fdk-container-distributions fdk-distributions-yellow"
    >
      <strong>
        Begrenset offentlighet
      </strong>
      <div>
        <DistributionFormat
          code="RESTRICTED"
          key="dataset-distribution-format0"
          text="application/json"
          type={null}
        />
        <DistributionFormat
          code="RESTRICTED"
          key="dataset-distribution-format1"
          text="xml"
          type={null}
        />
      </div>
      <div
        id="search-hit-sample"
      >
        Eksempeldata tilgjengelig
      </div>
    </div>
  </div>
</Link>
`;

exports[`should render SearchHitItem with no title, description, accessRights and distribution with no format correctly 1`] = `
<Link
  className="fdk-a-search-hit"
  replace={false}
  title="Datasett: undefined"
  to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
  <div
    className="fdk-container-search-hit"
  >
    <h2 />
    <div
      className="fdk-dataset-themes"
    >
      <span>
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Datasettet
        </span>
        Eies av
         
        <span
          className="fdk-strong-virksomhet"
        >
          Testetaten
        </span>
      </span>
    </div>
    <p
      className="fdk-p-search-hit"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Beskrivelse av datasettet,
      </span>
    </p>
    <div
      className=""
    >
      <strong />
      <div />
    </div>
  </div>
</Link>
`;
